webext permissions: activeTab and menu for a subframe

maxnikulin opened the pull request:
### Description

Clarify `aciveTab` permission effect on `executeScript` when context menu is opened for a subframe.

### Motivation

Developers may expect that acquiring the `activeTab` permission though add-on's context menu item gives access to the frame where the menu is opened, but actually content script may be injected into the top tab page.

My confusion led to creation of the #27098 wrong pull request.

maxnikulin committed:
webext permissions: activeTab and menu for a subframe

Explicitly state that `activeTab` gives permission for the tab's top level frame, so a particular frame, for which context menu is opened, may be inaccessible for add-on scripting due to various reasons: missed hosts permission, `data:` URI frames, etc.

Discussion in the Chromium bug tracker: <https://crbug.com/826433>

maxnikulin committed:
fixup! webext permissions: activeTab and menu for a subframe

Behavior of mv2 extensions in Firefox is more confusing in respect to cross-origin scripts and frames.

maxnikulin commented:
> Flaws (1)
>
>    macros:
>        /en-US/docs/Web/API/URL/createObjectURL redirects to /en-US/docs/Web/API/URL/createObjectURL_static

https://github.com/mdn/content/blob/d3b92dff059a74d95a6b1d64908b6992d57b21fd/files/en-us/web/api/blob/index.md?plain=1#L155 contains `{{DOMxRef("URL.createObjectURL")}}` as well. Should `{{DOMxRef("URL.createObjectURL()")}}` be used instead? Is the linter unhappy that I put "Blob" as link description?

> External URLs (1)
>
>    https://bugzil.la/1411641 (1 time) (Note! This may be a new URL eyes)

This page already contains a bugzil.la link, so I am surprised by this warning.

Perhaps added fragment is too heavy loaded with links. Should I remove some of them?

rebloor commented:
@Rob--W as this content appears to relate to a comment to left (https://bugzilla.mozilla.org/show_bug.cgi?id=1411641#c41) would you have a moment to check the technical accuracy before I look at the language?

rebloor commented:
@maxnikulin regarding the linter feedback. In both cases these are informational. The first one tells you that `{{DOMxRef("URL.createObjectURL")}}` passes through redirect to find the correct page. And the second is just advice that you are adding an external link.

Rob--W commented:
> Developers may expect that acquiring the `activeTab` permission though add-on's context menu item gives access to the frame where the menu is opened, but actually content script may be injected into the top tab page.

This is a reasonable expectation. I opened https://bugzilla.mozilla.org/show_bug.cgi?id=1838753 (and linked it to the other Bugzilla report you've filed before).

As for this documentation PR. There are several technically distinct parts that are grouped together in one large paragraph. I'll comment on them separately:

> You may face limitations of the `activeTab` permission trying to inject [content scripts](/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts) into web pages with {{HTMLElement("iframe", "frames")}}.

This piece of information is useful. But I would put the message near the existing bullet point, at https://github.com/mdn/content/blob/839d682cac996a347ea2422edc41ea73dfa73c92/files/en-us/mozilla/add-ons/webextensions/manifest.json/permissions/index.md?plain=1#L155

(side note: this does not mention the scripting API. For details, see "At runtime, into specific tabs" at https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts#loading_content_scripts )

> When the {{webextAPIref("menus/onClicked", "menus.onClicked")}} event is fired in response to a [context menu](/en-US/docs/Mozilla/Add-ons/WebExtensions/user_interface/Context_menu_items) opened inside a frame, the extension receives access to the **top-level frame** of the tab. As a result, the frame specified by the `frameId` field of {{webextAPIref("menus/OnClickData", "menus.OnClickData")}} may be inaccessible for injection of scripts and styles by {{webextAPIref("scripting/executeScript", "scripting.executeScript")}} and {{webextAPIref("scripting/insertCSS", "scripting.insertCSS")}} or by their {{webextAPIref("tabs", "tabs")}} counterparts.

This is specific to the activeTab permission & context menus API. This does not need a separate mention if we document that `activeTab` only grants access to the tab and same-origin frames.

> Running script in a [cross-origin](/en-US/docs/Web/Security/Same-origin_policy#cross-origin_network_access) frame may require [host permissions](#host_permissions). Support of special cases like [`data:` URI](/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URLs) or {{DOMxRef("URL.createObjectURL", "Blob")}} frames in respect to content scripts is not uniform across browsers, see the Firefox [bug #1411641 comment 41](https://bugzil.la/1411641#c41) for some details.

This has nothing to do with activeTab. This would fit better in the content scripts documentation, https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts#loading_content_scripts

maxnikulin commented:
> > Running script in a [cross-origin](/en-US/docs/Web/Security/Same-origin_policy#cross-origin_network_access) frame may require [host permissions](#host_permissions). Support of special cases like [`data:` URI](/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URLs) or {{DOMxRef("URL.createObjectURL", "Blob")}} frames in respect to content scripts is not uniform across browsers, see the Firefox [bug #1411641 comment 41](https://bugzil.la/1411641#c41) for some details.
> 
> This has nothing to do with activeTab. This would fit better in the content scripts documentation, https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts#loading_content_scripts

@Rob--W, I agree with this idea. My plan:

1. Create new section "Permissions, restrictions, and limitations"
   https://github.com/mdn/content/blob/839d682cac996a347ea2422edc41ea73dfa73c92/files/en-us/mozilla/add-ons/webextensions/content_scripts/index.md?plain=1#L43
    making "Restricted domains" its subsection.
2. Move notes (converting them to regular paragraphs) related to permission to first subsection of the new section
   https://github.com/mdn/content/blob/839d682cac996a347ea2422edc41ea73dfa73c92/files/en-us/mozilla/add-ons/webextensions/content_scripts/index.md?plain=1#L17
3. Add a link to "activeTab", mention the `scripting` permission.
4. Add a subsection for frames having special URIs based on the sentence cited above. I can add more links to browser bugs, but I am unsure if it is necessary.

> > You may face limitations of the `activeTab` permission trying to inject [content scripts](/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts) into web pages with {{HTMLElement("iframe", "frames")}}.
> 
> This piece of information is useful. But I would put the message near the existing bullet point, at
> 
> https://github.com/mdn/content/blob/839d682cac996a347ea2422edc41ea73dfa73c92/files/en-us/mozilla/add-ons/webextensions/manifest.json/permissions/index.md?plain=1#L155

I would not insist, however originally I decided to keep the bullet list concise and to provide details in discussion paragraphs below.

> (side note: this does not mention the scripting API. For details, see "At runtime, into specific tabs" at https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts#loading_content_scripts )

The "Host permission" section has the same issue, but expecting possible objections related to context menu case I decided to keep changes minimal.

> > When the {{webextAPIref("menus/onClicked", "menus.onClicked")}} event is fired in response to a [context menu](/en-US/docs/Mozilla/Add-ons/WebExtensions/user_interface/Context_menu_items) opened inside a frame, the extension receives access to the **top-level frame** of the tab. As a result, the frame specified by the `frameId` field of {{webextAPIref("menus/OnClickData", "menus.OnClickData")}} may be inaccessible for injection of scripts and styles by {{webextAPIref("scripting/executeScript", "scripting.executeScript")}} and {{webextAPIref("scripting/insertCSS", "scripting.insertCSS")}} or by their {{webextAPIref("tabs", "tabs")}} counterparts.
> 
> This is specific to the activeTab permission & context menus API. This does not need a separate mention if we document that `activeTab` only grants access to the tab and same-origin frames.

My idea is to stress on the case of cross-origin context menu either here in permissions or in the `menus.onClicked` docs. An alternative (links are omitted):

The `acitveTab` permission enables scripting access to the top level tab's frame and same origin frames. Running scripts or adding styles to cross-origin frames may require additional host permissions. It includes a case when a context menu is opened inside a frame. Even the frame specified as `frameId` in the argument of a `menus.onClicked` event listener may be inaccessible. The intention is to avoid escalation of privileges by creating a frame with arbitrary web page and running some action in its context. Limitations imposed by browsers are not uniform, for example Firefox blocks access to frames created by page scripts. On the other hand only Firefox mv2 add-ons are allowed to run context scripts in static cross-origin frames. Of course, restrictions and limitations related to particular sites and URI schemes applies as well.

Should I add "Content scripts" edits to this pull request or it is better to create a dedicated one?

Rob--W commented:
Feel free to update this PR with what you have in mind. Your objective is not tied to one article, and having one PR would make it easier to get the message to the right audience when needed.

Try to be clear and concise, so that the main message doesn't get lost in the sea of words.

maxnikulin committed:
webext content scripts: group permissions and restricted domains

- A new section to combine description of permissions
  and exceptions for some sites
- Move notes on permissions to a subsection of the new
  section
- Add links to `activeTab` and `scripting` in the context
  of runtime script injection.

maxnikulin committed:
webext content scripts: data: and blob: may be inaccessible

Depending on browser, documents having special URI schemes may be inaccessible for extensions.

maxnikulin committed:
webext permissions: mention scripting along with tabs

Add `scripting.executeScript` and `scripting.insertCSS`
to the `tabs` counterparts in description hosts and `activeTab`
permissions.

rebloor commented on files/en-us/mozilla/add-ons/webextensions/manifest.json/permissions/index.md:
```diff
@@ -56,7 +56,9 @@ The extra privileges include:
 - the ability to read tab-specific metadata without the "tabs" permission, such as the `url`, `title`, and `favIconUrl` properties of {{WebExtAPIRef("tabs.Tab")}} objects
-- the ability to inject scripts programmatically (using {{webextAPIref("tabs/executeScript", "tabs.executeScript()")}}) into pages served from those origins
+- the ability to inject scripts and styles programmatically into pages served from those origins
+  -  using {{webextAPIref("scripting/executeScript", "scripting.executeScript()")}} and {{webextAPIref("scripting/insertCSS", "scripting.insertCSS()")}} when combined with the `scripting` permission (in Chrome mv3 only)
```
```suggestion
  - using {{webextAPIref("scripting/executeScript", "scripting.executeScript()")}} and {{webextAPIref("scripting/insertCSS", "scripting.insertCSS()")}} when combined with the `scripting` permission (in Chrome Manifest V3 only.)
```

rebloor commented on files/en-us/mozilla/add-ons/webextensions/manifest.json/permissions/index.md:
```diff
@@ -56,7 +56,9 @@ The extra privileges include:
-- the ability to inject scripts programmatically (using {{webextAPIref("tabs/executeScript", "tabs.executeScript()")}}) into pages served from those origins
+- the ability to inject scripts and styles programmatically into pages served from those origins
+  -  using {{webextAPIref("scripting/executeScript", "scripting.executeScript()")}} and {{webextAPIref("scripting/insertCSS", "scripting.insertCSS()")}} when combined with the `scripting` permission (in Chrome mv3 only)
+  -  using {{webextAPIref("tabs/executeScript", "tabs.executeScript()")}} and {{webextAPIref("tabs/insertCSS", "tabs.insertCss()")}} (mv2 only)
```
```suggestion
  - using {{webextAPIref("tabs/executeScript", "tabs.executeScript()")}} and {{webextAPIref("tabs/insertCSS", "tabs.insertCss()")}}. (Manifest V2 only)
```

maxnikulin replied:
Is it OK to insert full stop if the middle of an item without capitalizing first letter and without adding another dot at the end? I am unsure concerning proper punctuation. Should not all list items be rewritten using complete sentences?

rebloor commented on files/en-us/mozilla/add-ons/webextensions/manifest.json/permissions/index.md:
```diff
@@ -152,7 +154,9 @@ This permission is specified as `"activeTab"`. If an extension has the `activeTa
 
-- The ability to inject JavaScript or CSS into the tab programmatically, using {{webextAPIref("tabs/executeScript", "browser.tabs.executeScript()")}} and {{webextAPIref("tabs/insertCSS", "browser.tabs.insertCSS()")}}
+- The ability to inject JavaScript or CSS into the tab programmatically
+  -  using {{webextAPIref("scripting/executeScript", "scripting.executeScript()")}} and {{webextAPIref("scripting/insertCSS", "scripting.insertCSS()")}} when combined with the `scripting` permission (in Chrome mv3 only)
```
```suggestion
  - using {{webextAPIref("scripting/executeScript", "scripting.executeScript()")}} and {{webextAPIref("scripting/insertCSS", "scripting.insertCSS()")}} when combined with the `scripting` permission (in Chrome Manifest V3 only.)
```

rebloor commented on files/en-us/mozilla/add-ons/webextensions/manifest.json/permissions/index.md:
```diff
@@ -152,7 +154,9 @@ This permission is specified as `"activeTab"`. If an extension has the `activeTa
-- The ability to inject JavaScript or CSS into the tab programmatically, using {{webextAPIref("tabs/executeScript", "browser.tabs.executeScript()")}} and {{webextAPIref("tabs/insertCSS", "browser.tabs.insertCSS()")}}
+- The ability to inject JavaScript or CSS into the tab programmatically
+  -  using {{webextAPIref("scripting/executeScript", "scripting.executeScript()")}} and {{webextAPIref("scripting/insertCSS", "scripting.insertCSS()")}} when combined with the `scripting` permission (in Chrome mv3 only)
+  -  using {{webextAPIref("tabs/executeScript", "tabs.executeScript()")}} and {{webextAPIref("tabs/insertCSS", "tabs.insertCss()")}} (mv2 only)
```
```suggestion
  - using {{webextAPIref("tabs/executeScript", "tabs.executeScript()")}} and {{webextAPIref("tabs/insertCSS", "tabs.insertCss()")}}. (Manifest V2 only)
```

rebloor commented on files/en-us/mozilla/add-ons/webextensions/manifest.json/permissions/index.md:
```diff
@@ -161,6 +165,8 @@ For example, consider an extension that wants to run a script in the current pag
 
 > **Note:** You can only get access to the tab/data that was there, when the user interaction occurred (e.g. the click). When the active tab navigates away (e.g., due to finishing loading or some other event), the permission does not grant you access to the tab anymore.
 
+You may face limitations of the `activeTab` permission trying to inject [content scripts](/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts) into web pages with {{HTMLElement("iframe", "frames")}}. When the {{webextAPIref("menus/onClicked", "menus.onClicked")}} event is fired in response to a [context menu](/en-US/docs/Mozilla/Add-ons/WebExtensions/user_interface/Context_menu_items) opene ...
```
```suggestion
You may encounter limitations with the `activeTab` permission when injecting [content scripts](/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts) into web pages with {{HTMLElement("iframe", "frames")}}. When the {{webextAPIref("menus/onClicked", "menus.onClicked")}} event is fired in response to a [context menu](/en-US/docs/Mozilla/Add-ons/WebExtensions/user_interface/Context_menu_items) opened inside a frame, the extension receives access to the **top-level frame** of the tab. As a result, the frame specified by the `frameId` field of {{webextAPIref("menus/OnClickData", "menus.OnClickData")}} may be inaccessible for injection of scripts and styles by {{webextAPIref("scripting/executeScript", "scripting.executeScript")}} and {{webextAPIref("scripting/insertCSS", "scripting.insertCSS")}} or by their {{webextAPIref("tabs", "tabs")}} counterparts. Running script in a [cross-origin](/en-US/docs/Web/Security/Same-origin_policy#cross-origin_network_access) frame may require [host permissions](#host_permissions). Support of special cases like [`data:` URI](/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URLs) or {{DOMxRef("URL.createObjectURL", "Blob")}} frames in respect to content scripts is not uniform across browsers, see the Firefox [bug #1411641 comment 41](https://bugzil.la/1411641#c41) for some details.
```

maxnikulin replied:
I am sorry, I did not finish edits yesterday. The final part has been moved to the article on content script. I am still in doubts if the remaining part should be moved above into the list describing the privileges. For while I have tried to make it a bit shorter. The updated variant is based on my https://github.com/mdn/content/pull/27219#issuecomment-1596986757

maxnikulin replied:
I do not mind to revert last change and to take your variant

rebloor reviewed (commented):
Some suggestions for your consideration

maxnikulin committed:
fixup! webext permissions: activeTab and menu for a subframe

Rewrite the note on `activeTab` limitation in respect to cross-origin frames

maxnikulin committed:
webext permissions: action and browser action in activeTab

Update user interaction variants in the `activeTab` description to reflect changed in Manifest V3.

maxnikulin committed:
fixup! webext permissions: mention scripting along with tabs

- Fix linter errors (double spaces after dash)
- Use "Manifest V2/V3" instead of "mv2/mv3"

maxnikulin commented on files/en-us/mozilla/add-ons/webextensions/manifest.json/permissions/index.md:
```diff
@@ -161,6 +166,8 @@ For example, consider an extension that wants to run a script in the current pag
 
 > **Note:** You can only get access to the tab/data that was there, when the user interaction occurred (e.g. the click). When the active tab navigates away (e.g., due to finishing loading or some other event), the permission does not grant you access to the tab anymore.
 
+The `acitveTab` permission enables scripting access to the top level tab's page and same origin frames. Running scripts or modifying styles inside [cross-origin](/en-US/docs/Web/Security/Same-origin_policy#cross-origin_network_access) frames may require additional [host permissions](#host_permissions). Even a listener of the {{webextAPIref("menus/onClicked", "menus.onClicked")}} event may be unab ...
```
@Rob--W, I may try to move this paragraph to make it a part of the list item above. It would be easy to revert such change.

Rob--W commented on files/en-us/mozilla/add-ons/webextensions/content_scripts/index.md:
```diff
@@ -14,12 +14,6 @@ Just like the scripts loaded by normal web pages, content scripts can read and m
 
 Content scripts can only access [a small subset of the WebExtension APIs](#webextension_apis), but they can [communicate with background scripts](#communicating_with_background_scripts) using a messaging system, and thereby indirectly access the WebExtension APIs.
 
-> **Note:** Content scripts are only executed if the extension is granted [host permissions](/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#host_permissions) for the domain.
```
With this removed, it may still be worth keeping a reference to the new section, to emphasize that scripts are only executed if access is granted.

rebloor committed:
Apply suggestions from review

Co-authored-by: Rob Wu <rob@robwu.nl>

rebloor commented on files/en-us/mozilla/add-ons/webextensions/content_scripts/index.md:
```diff
@@ -14,12 +14,6 @@ Just like the scripts loaded by normal web pages, content scripts can read and m
 
```
@Rob--W did you mean something like this:
```suggestion
However, they can only do this when [granted host permission](#permissions).
```

Rob--W replied:
This looks good.

rebloor committed:
Linter suggestions

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

rebloor committed:
Apply suggestions from review and linter

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

rebloor committed:
Merge branch 'main' into activetab-frame-2

rebloor committed:
Tidy up load options list format

rebloor committed:
Merge branch 'main' into activetab-frame-2

Rob--W commented on files/en-us/mozilla/add-ons/webextensions/content_scripts/index.md:
```diff
@@ -10,30 +10,20 @@ A content script is a part of your extension that runs in the context of a parti
 [Background scripts](/en-US/docs/Mozilla/Add-ons/WebExtensions/Background_scripts) can access all the [WebExtension JavaScript APIs](/en-US/docs/Mozilla/Add-ons/WebExtensions/API), but they can't directly access the content of web pages. So if your extension needs to do that, you need content scripts.
 
-Just like the scripts loaded by normal web pages, content scripts can read and modify the content of their pages using the standard DOM APIs.
+Just like the scripts loaded by normal web pages, content scripts can read and modify the content of their pages using the standard DOM APIs. However, they can only do this when [granted host permission](#permissions).
```
"when granted host permission" - is that proper English? At my first review pass I read "permissions" (plural), but I just noticed that the text label itself is singular. How about this:

```suggestion
Just like the scripts loaded by normal web pages, content scripts can read and modify the content of their pages using the standard DOM APIs. However, they can only do this when [host permissions to the web page's origin have been granted](#permissions).
```

rebloor committed:
Review feedback

Co-authored-by: Rob Wu <rob@robwu.nl>

rebloor commented:
Thanks for your help @maxnikulin
